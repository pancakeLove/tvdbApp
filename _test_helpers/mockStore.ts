import configureMockStore from 'redux-mock-store';
import createSagaMiddleware from 'redux-saga';

export const mockStore = (state: Object, tvdbApi: any, saga: any) => {
  const context = {
    tvdbApi,
  };
  const sagaMiddleware = createSagaMiddleware();
  sagaMiddleware.setContext(context);
  const mockStore = configureMockStore([sagaMiddleware]);
  const store = mockStore(state);
  sagaMiddleware.run(saga);
  return {...store};
};
