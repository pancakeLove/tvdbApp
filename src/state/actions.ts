import {createStandardAction} from 'typesafe-actions';
import {SearchResultItem, SeriesDetails} from '../api/tvdb/types';
import {GeneralError} from '../shared/types';

export enum TvDbActionTypes {
  LOGIN_REQUEST = 'LOGIN_REQUEST',
  LOGIN_SUCCESS = 'LOGIN_SUCCESS',
  LOGIN_FAILURE = 'LOGIN_FAILURE',
  SEARCH_REQUEST = 'SEARCH_REQUEST',
  SEARCH_SUCCESS = 'SEARCH_SUCCESS',
  DETAILS_REQUEST = 'DETAILS_REQUEST',
  DETAILS_SUCCESS = 'DETAILS_SUCCESS',
  API_ERROR = 'API_ERROR',
  IS_LOADING = 'IS_LOADING',
}

export const tvDbActions = {
  loginRequest: createStandardAction(TvDbActionTypes.LOGIN_REQUEST)<null>(),
  loginSuccess: createStandardAction(TvDbActionTypes.LOGIN_SUCCESS)<string>(),
  loginFailure: createStandardAction(TvDbActionTypes.LOGIN_FAILURE)(),
  searchRequest: createStandardAction(TvDbActionTypes.SEARCH_REQUEST)<string>(),
  searchSuccess: createStandardAction(TvDbActionTypes.SEARCH_SUCCESS)<
    SearchResultItem[]
  >(),
  detailsRequest: createStandardAction(
    TvDbActionTypes.DETAILS_REQUEST,
  )<number>(),
  detailsSuccess: createStandardAction(
    TvDbActionTypes.DETAILS_SUCCESS,
  )<SeriesDetails | null>(),
  apiError: createStandardAction(
    TvDbActionTypes.API_ERROR,
  )<GeneralError | null>(),
  isLoading: createStandardAction(TvDbActionTypes.IS_LOADING)<boolean>(),
};
