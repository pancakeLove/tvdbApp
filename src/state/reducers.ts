import {Reducer} from 'redux';
import {TvDbActionTypes} from './actions';
import {Action} from 'typesafe-actions';
import {SearchResultItem, SeriesDetails} from '../api/tvdb/types';
import {GeneralError} from '../shared/types';

export interface TvDbState {
  loginInit: boolean;
  token: string;
  loginError: boolean;
  searchResult?: SearchResultItem[];
  apiError?: GeneralError;
  isLoading: boolean;
  seriesDetails?: SeriesDetails;
}

const initialState = {
  loginInit: false,
  token: '',
  loginError: false,
  searchResult: [],
  apiError: {error: false, message: ''},
  isLoading: false,
};

export interface DispatchAction extends Action<TvDbActionTypes> {
  payload: Partial<TvDbState>;
}

export const tvDbReducer: Reducer<TvDbState, DispatchAction> = (
  state = initialState,
  action,
) => {
  switch (action?.type) {
    case TvDbActionTypes.LOGIN_REQUEST:
      return {...state, loginError: false, loginInit: true};
    case TvDbActionTypes.LOGIN_SUCCESS:
      return {...state, token: action.payload as string, isLoading: false};
    case TvDbActionTypes.LOGIN_FAILURE:
      return {
        ...state,
        token: '',
        loginError: true,
      };
    case TvDbActionTypes.SEARCH_SUCCESS:
      return {
        ...state,
        searchResult: action.payload as SearchResultItem[],
        isLoading: false,
      };
    case TvDbActionTypes.DETAILS_SUCCESS:
      return {
        ...state,
        seriesDetails: action.payload as SeriesDetails,
        isLoading: false,
      };
    case TvDbActionTypes.API_ERROR:
      return {
        ...state,
        apiError: action.payload as GeneralError,
        isLoading: false,
      };
    default:
      return state;
  }
};
