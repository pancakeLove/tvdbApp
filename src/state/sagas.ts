import {
  takeLatest,
  call,
  put,
  all,
  select,
  getContext,
} from 'redux-saga/effects';
import {tvDbActions, TvDbActionTypes} from './actions';
import {
  LoginResponse,
  SearchResponse,
  SeriesDetailsResponse,
} from '../api/tvdb/types';
import {AxiosError} from 'axios';
import {TvDbState} from './reducers';
import {PayloadAction} from 'typesafe-actions';

function* handleApiError(error: AxiosError<any>) {
  let message: string = '';
  if (error.response) {
    message = 'Server responded with non 2xx status: ' + error.response.status;
  } else if (error.request) {
    message = 'Timeout received, possible slow network';
  } else {
    message = error.message;
  }
  yield put(tvDbActions.apiError({error: true, message}));
}

export function* loginWatcher() {
  yield takeLatest(TvDbActionTypes.LOGIN_REQUEST, loginWorker);
}
export function* loginWorker() {
  yield put(tvDbActions.isLoading(true));
  const tvdbApi = yield getContext('tvdbApi');
  //console.log('tvdbApi from context:', tvdbApi);
  const response: LoginResponse = yield call(tvdbApi().fetchToken);
  if (response.error) {
    yield put({type: TvDbActionTypes.LOGIN_FAILURE});
    yield handleApiError(response.error);
    yield put(tvDbActions.isLoading(false));
  } else {
    yield put(tvDbActions.loginSuccess(response.token));
  }
}

export function* searchWatcher() {
  yield takeLatest(TvDbActionTypes.SEARCH_REQUEST, searchWorker);
}
function* searchWorker(action: PayloadAction<any, string>) {
  yield put(tvDbActions.isLoading(true));
  const tvdbApi = yield getContext('tvdbApi');
  const token = yield select((state: TvDbState) => state.token);
  const response: SearchResponse = yield call(
    tvdbApi().search,
    action.payload,
    token,
  );
  if (response.error) {
    yield handleApiError(response.error);
    yield put(tvDbActions.isLoading(false));
  } else {
    /* todo: should be some limitation to number of results, eg:
     yield put(tvDbActions.searchSuccess(response.data?.slice(0, 100)));
     */
    yield put(tvDbActions.searchSuccess(response.data));
  }
}

export function* fetchDetailsWatcher() {
  yield takeLatest(TvDbActionTypes.DETAILS_REQUEST, fetchDetailsWorker);
}
function* fetchDetailsWorker(action: PayloadAction<any, number>) {
  yield put(tvDbActions.isLoading(true));
  const tvdbApi = yield getContext('tvdbApi');
  const token = yield select((state: TvDbState) => state.token);
  const response: SeriesDetailsResponse = yield call(
    tvdbApi().details,
    action.payload,
    token,
  );
  if (response.error) {
    yield handleApiError(response.error);
    yield put(tvDbActions.isLoading(false));
  } else {
    yield put(tvDbActions.detailsSuccess(response.data));
  }
}

export function* tvDbSagas() {
  yield all([loginWatcher(), searchWatcher(), fetchDetailsWatcher()]);
}
