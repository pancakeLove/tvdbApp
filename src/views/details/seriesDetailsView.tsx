import React, {useEffect} from 'react';
import {
  StyleSheet,
  View,
  Text,
  Modal,
  ScrollView,
  Button,
  ActivityIndicator,
} from 'react-native';
import {useDispatch, useSelector} from 'react-redux';
import {TvDbState} from '../../state/reducers';
import {tvDbActions} from '../../state/actions';
import {actionColor} from '../../shared/styles';

export interface Props {
  readonly seriesId: number;
  readonly onClose: any;
  readonly visible: boolean;
}

export const SeriesDetails = (props: Props) => {
  const dispatch = useDispatch();
  const {seriesDetails, isLoading} = useSelector((s: TvDbState) => s);
  useEffect(() => {
    dispatch(tvDbActions.detailsRequest(props.seriesId));
  }, [props.seriesId]);
  return (
    <Modal
      visible={props.visible}
      animationType={'slide'}
      presentationStyle={'overFullScreen'}
      transparent={true}>
      {isLoading ? (
        <ActivityIndicator
          size="large"
          color="white"
          style={{flex: 1, alignSelf: 'center'}}
        />
      ) : (
        <ScrollView contentContainerStyle={styles.fullscreen}>
          <Text style={styles.title}>{seriesDetails?.seriesName}</Text>
          <Text style={styles.text}>{seriesDetails?.rating}</Text>
          <Text style={styles.text}>{seriesDetails?.overview}</Text>
          <Button
            onPress={() => props.onClose()}
            title="Close"
            color={actionColor}
            accessibilityLabel="Close"
          />
        </ScrollView>
      )}
    </Modal>
  );
};

const styles = StyleSheet.create({
  text: {
    width: '90%',
    color: 'lightgray',
    padding: 10,
    alignSelf: 'center',
    borderColor: 'gray',
    borderWidth: 1,
  },
  title: {
    color: 'white',
    fontSize: 18,
    margin: 8,
    alignSelf: 'center',
  },
  fullscreen: {
    flex: 1,
    backgroundColor: 'black',
    alignContent: 'center',
    justifyContent: 'center',
  },
});
