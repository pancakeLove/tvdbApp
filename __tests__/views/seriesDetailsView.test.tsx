import React from 'react';
import {create} from 'react-test-renderer';
import {Provider} from 'react-redux';
import {mockStore} from '../../_test_helpers/mockStore';
import {tvDbReducer} from '../../src/state/reducers';
import {tvDbSagas} from '../../src/state/sagas';
import {SeriesDetails as SeriesDetailsView} from '../../src/views/details/seriesDetailsView';
import {
  LoginResponse,
  SearchResponse,
  SeriesDetails,
  SeriesDetailsResponse,
} from '../../src/api/tvdb/types';
import {Button} from 'react-native';

describe('DetailsView', async () => {
  const detailsSuccessResponse: SeriesDetailsResponse = {
    data: {} as SeriesDetails,
    error: undefined,
  };
  const loginSuccessResponse: LoginResponse = {
    token: 'token',
    error: undefined,
  };
  const searchSuccessResponse: SearchResponse = {
    data: {} as any,
    error: undefined,
  };
  const tvdbApi = () => ({
    details: jest.fn().mockReturnValue(detailsSuccessResponse),
    fetchToken: jest.fn().mockReturnValue(loginSuccessResponse),
    search: jest.fn().mockReturnValue(searchSuccessResponse),
  });
  const store = mockStore(tvDbReducer, tvdbApi, tvDbSagas);
  const onCloseCallback = jest.fn();

  await test('testing close button', async () => {
    let tree = create(
      <Provider store={store}>
        <SeriesDetailsView
          seriesId={1}
          visible={true}
          onClose={onCloseCallback}
        />
      </Provider>,
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
