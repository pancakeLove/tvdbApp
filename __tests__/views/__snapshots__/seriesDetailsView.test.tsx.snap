// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailsView testing close button 1`] = `
<Modal
  animationType="slide"
  hardwareAccelerated={false}
  presentationStyle="overFullScreen"
  transparent={true}
  visible={true}
>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "alignContent": "center",
        "backgroundColor": "black",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <View>
      <Text
        style={
          Object {
            "alignSelf": "center",
            "color": "white",
            "fontSize": 18,
            "margin": 8,
          }
        }
      />
      <Text
        style={
          Object {
            "alignSelf": "center",
            "borderColor": "gray",
            "borderWidth": 1,
            "color": "lightgray",
            "padding": 10,
            "width": "90%",
          }
        }
      />
      <Text
        style={
          Object {
            "alignSelf": "center",
            "borderColor": "gray",
            "borderWidth": 1,
            "color": "lightgray",
            "padding": 10,
            "width": "90%",
          }
        }
      />
      <View
        accessibilityLabel="Close"
        accessibilityRole="button"
        accessibilityState={Object {}}
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
                Object {
                  "color": "#841584",
                },
              ]
            }
          >
            Close
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</Modal>
`;
